### PR TITLE
STITCH-1250 - Implement FCM push and service client in Core, Android, and Server

### PR DESCRIPTION
**Pull request by edaniels**
No push client on server.

cc: @jsflax no review needed just and FYI if you want to look

### CODE REVIEW

**Comment by edaniels** [comment]
@adamchel @jsflax you can see the usage in the example app built (fcm)

File: core/sdk/src/main/java/com/mongodb/stitch/core/push/internal/CoreStitchPushClient.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+package com.mongodb.stitch.core.push.internal;
+
+import org.bson.Document;
+
+public interface CoreStitchPushClient {
+  void registerInternal(final Document registrationInfo);
+
+  void deregisterInternal();
+}
```
Comment (adamchel) [inline core/sdk/src/main/java/com/mongodb/stitch/core/push/internal/CoreStitchPushClient.java]: [nit] newline

File: core/services/fcm/build.gradle [hunk 2]
```diff
@@ -0,0 +1,22 @@
+apply plugin: 'java-library'
+apply plugin: 'com.jfrog.bintray'
+
+ext.pomDisplayName = "Core Twilio Service"
```
Comment (adamchel) [inline core/services/fcm/build.gradle]: [nit] FCM

File: server/core/src/main/java/com/mongodb/stitch/server/core/services/internal/StitchServiceClientImpl.java [hunk 3]
```diff
@@ -58,10 +59,10 @@ public StitchServiceImpl(
...
-          final Long requestTimeout,
-          final Decoder<ResultT> resultDecoder) {
+      final String name,
+      final List<?> args,
+      final Long requestTimeout,
+      final Decoder<ResultT> resultDecoder) {
     return callFunctionInternal(name, args, requestTimeout, resultDecoder);
   }
-}
+}
```
Comment (adamchel) [inline server/core/src/main/java/com/mongodb/stitch/server/core/services/internal/StitchServiceClientImpl.java]: [nit] newline

File: core/services/fcm/src/main/java/com/mongodb/stitch/core/services/fcm/FcmSendMessageRequest.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+   * @return the group of messages that can be collapsed.
+   */
+  @Nullable
+  public String getCollapseKey() {
+    return collapseKey;
+  }
+
+  /**
+   * Returns whether or not to indicate to the client that content is available in order
+   * to wake the device. Note: iOS only.
```
Comment (adamchel) [inline core/services/fcm/src/main/java/com/mongodb/stitch/core/services/fcm/FcmSendMessageRequest.java]: [nit] "iOS only" could be a little confusing. Maybe "for messages to iOS devices only". Also applies to other places in this file, and there were also places that said "Android only" that could be a little more clear

File: android/core/src/main/java/com/mongodb/stitch/android/core/push/internal/StitchPushClient.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package com.mongodb.stitch.android.core.push.internal;
+
+import com.google.android.gms.tasks.Task;
+import com.mongodb.stitch.core.push.internal.CoreStitchPushClient;
+
+import org.bson.Document;
+
+public interface StitchPushClient extends CoreStitchPushClient {
```
[the remark this answers is not in the document]
Comment (edaniels) [reply android/core/src/main/java/com/mongodb/stitch/android/core/push/internal/StitchPushClient.java]: This is an internal class that shouldn't be used externally

File: core/services/fcm/src/main/java/com/mongodb/stitch/core/services/fcm/FcmSendMessageRequest.java (hunk 4 above)
Comment (edaniels) [reply core/services/fcm/src/main/java/com/mongodb/stitch/core/services/fcm/FcmSendMessageRequest.java]: Ok

**Pull request merged**